### DISCUSSION TITLE
METviewer PCT statistics

### DISCUSSION THREAD

**Discussion by lindsayrblank** [dtcenter/METplus#1448] [Plot Generation] [answered]
Hello all! 

I am using METviewer v4.0.0 running on an AWS instance. I have uploaded PCT data to the database. When I go to calculate statistics, the only option is 'PCT_TOTAL'. Unlike CTC or SL1L2, the options of all the statistics that can be calculated from PCT, like PSTD, are not listed, yet PCT is an option in the statistics box dropdown menu. Is this expected behavior?

Best,
Lindsay

**Comment by TatianaBurek** [comment, marked as answer]
Lindsay
Which instance of AWS with METviewer are you talking about? is it NOAA/NCAR or something else? Can I see the web page?
By "go to calculate statistics" do you mean "Dependent (Forecast) Variables" section with the dropdown list of stats?
This list is created bases on the data in the database tables. PCT statistics (PSTD_BASER, PSTD_RELIABILITY...) are only present if there is data in line_data_pstd table. 
We don't have other statistics that start from PCT except PCT_TOTAL.

**Reply by lindsayrblank** [reply]
HI Tatiana, 

The dropdown list of stats is what I am talking about. Okay, that is the behavior I am seeing. Is there a reason why probabilistic statistics can't be derived from PCTs like categorical statistics can be derived from CTCs? Is this something under development?

Best,
Lindsay

**Reply by JohnHalleyGotway** [reply]
Lindsay, I went hunting around for an example to see if I could understand this situation better.

At this [METviewer instance](http://dtcenter.ucar.edu/met/metviewer/metviewer1.jsp), I created this plot and XML:
[plot_20220225_211909.xml.txt](https://github.com/dtcenter/METplus/files/8144917/plot_20220225_211909.xml.txt)
![plot_20220225_211909](https://user-images.githubusercontent.com/21087144/155804718-d2c0379f-d279-4a21-8b67-e7b7ee054483.png)

When making this plot, I DO see the PCT_TOTAL statistic in the list followed by several PSTD_* statistic options.
<img width="311" alt="Screen Shot 2022-02-25 at 2 22 53 PM" src="https://user-images.githubusercontent.com/21087144/155804787-13c6606d-3233-4f67-9767-61b91889ef9e.png">

I suspect that this database contains both PCT and PSTD output lines already. But I suspect that you're working with a database that ONLY contains PCT output lines and no PSTD ones.

I hear @TatianaBurek saying that those PSTD stats only show up in the list if the database table containing PSTD stats is non-empty. And I hear @lindsayrblank saying my database only contains PCT output, I want to be able to DERIVE AND PLOT the corresponding PSTD statistics (and perhaps PJC and PRC stats as well?).

@lindsayrblank, a temporary bandaid solution would be loading at least one .stat file with at least one PSTD line into your database. The hope is that that will enable the PSTD stats to show up in the drop-down list.

@TatianaBurek, I agree that it seems like we need an issue for refining the logic to decide which stats should show up in the dropdown list. Instead of only considering what's already present in the database, we should also consider what can be derived from what's present.

This logic should be roughly consistent with the "[aggregate_stat](https://met.readthedocs.io/en/develop/Users_Guide/stat-analysis.html#aggregate-stat-lines-and-produce-aggregated-statistics)" job type in Stat-Analysis. This is an incomplete list, but we'd start with something like this:
- Display CTS stats if the database contains CTS, CTC, or FHO data.
- Display MCTS stats if the database contains MCTS or MCTC data.
- Display CNT stats if the database contains CNT, SL1L2, or SAL1L2 data.
- Display VCNT stats if the database contains VCNT, VL1L2, or VAL1L2 data.
- Display PSTD stats if the database contains PSTD or PCT data.
- And so on.

@TatianaBurek if you agree with this logic, I'd recommend writing up a GitHub issue to capture the feature request. We could iterate on it to define exactly what the logic should be line type by line type.

**Reply by lindsayrblank** [reply]
Hi @JohnHalleyGotway and @TatianaBurek,

John, you nailed it. That's exactly what I want, a way to derive and plot. That's currently how I'm using METviewer to derive CNT and CTS statistics from SL1L2 and CTC line types respectively. The database I'm deriving those from does not include any CNT or CTS line types. I will try to add a line with a PSTD line type in it and see what happens. That will most likely happen later this week, but I'll let you know if I'm successful. 

Thank you both!

**Comment by TatianaBurek** [comment]
I created an issue in METviever repository:
https://github.com/dtcenter/METviewer/issues/379

**Reply by JohnHalleyGotway** [reply]
Thanks @TatianaBurek for writing up that issue. @lindsayrblank have you had a chance to test the work-around, loading at least some PSTD data into the database to trigger METviewer to serve up the PSTD stats?

Once we get this discussion wrapped up, I'd like to select one of the responses as being the best "Answer" and then "Lock conversation" on the discussion.

**Reply by lindsayrblank** [reply]
Great, thank you, Tatiana! I just tested @JohnHalleyGotway hypothesis of loading a database with both PCT and PSTD line types. It did not work. I loaded a PSTD line type with an initialization hour of 00 and two PCT files, one with an init hour of 00 and one with an init hour of 12. I could calculate PSTD statistics for the 00 hour only. It was not available for the 12. It was a good idea!

**Reply by JohnHalleyGotway** [reply]
@lindsayrblank FYI, METviewer [version 4.1.0](https://github.com/dtcenter/METviewer/releases/tag/v4.1.0) was released last Friday and includes [this fix](https://github.com/dtcenter/METviewer/issues/379) for that issue.

If you're able to test with the updated version of METviewer, please confirm that this issue has been resolved.

**Reply by lindsayrblank** [reply]
Hi John, thank you so much. I will get that working and let you know the results of the test.

**Reply by JohnHalleyGotway** [reply]
@lindsayrblank I'd like to go ahead and mark this discussion as being answered and then lock it to prevent future posts.

I'll plan to do that tomorrow unless I hear otherwise from you.

**Reply by lindsayrblank** [reply]
Hi John, 

I just got back from vacation and version 4.1.0 was installed while I was out. I will test it today and get back to you.

**Reply by lindsayrblank** [reply]
Hi @JohnHalleyGotway @TatianaBurek,

I'm trying to test the new changes, but for some reason, none of my databases show any available attributes. The SE (@greenlaw) who installed this version of METviewer said that he's seeing errors that look like this:

`2022-03-29 21:19 edu.ucar.metviewer.db.mysql.MysqlAppDatabaseManager.getListStat(Unknown Source) -1 getListStat - (conn=15) Table 'mv_grid_to_point.line_data_ssidx' doesn't exist`

What does this error mean and how do we fix it?

**Reply by JohnHalleyGotway** [reply]
@TatianaBurek is out of the office until 3/31, and we should wait for her expert opinion on this.

But here's my guess. Each new version of METviewer is usually accompanied with changes the database schema. Whenever we add new columns or line types to the output of MET, the METviewer schema is updated to know how to handle them. So any existing databases from the older version need to be updated to comply with the newer schema version.

Tatiana writes a database upgrade script for each new version. After installing the new METviewer version on a server, she runs that upgrade script on existing databases. And I assume she's done so for METviewer 4.1.0 as well. But I've never run that upgrade script myself and wouldn't know what to tell you to do.

Seems to me that those instructions belong in the METviewer User's Guide. Perhaps in the ["Installation Chapter"](https://metviewer.readthedocs.io/en/latest/Users_Guide/installation.html)? You might make a good guinea pig for the development and testing of those instructions!

**Reply by TatianaBurek** [reply]
Lindsay,
John gave you a correct answer. In addition to installing a new version of METviewer you need to run a script that updates the schema of all databases that are used with METviewer. I attached the script to this comment.
Please also notice that v4.1 uses Python as a default implementation for the statistics calculation and plotting. Rscript is deprecated and will be removed in the next release. If you wish to continue to use Rscript for now you need to explicitly specify it.
[update_for_4_1.sql.txt](https://github.com/dtcenter/METplus/files/8392986/update_for_4_1.sql.txt)

**Reply by lindsayrblank** [reply]
@TatianaBurek and @JohnHalleyGotway Thank you both for the information. We will be able to test the script in the next day or so and will report back the results.

**Reply by lindsayrblank** [reply]
Alright, we were able to run the SQL script and I was able to calculate PSTD statistics from a PCT line type. Thank you so much for all of your work on this issue. This functionality will really help!

**Reply by JohnHalleyGotway** [reply]
Thanks for confirming. I'll select and answer on this discussion and lock conversation on it.
